### PR TITLE
added proxy assess in feature space

### PR DESCRIPTION
**Pull request by amrasekh**
## Change description
Added proxy assessment in feature space

### CODE REVIEW

File: credoai/modules/dataset_modules/dataset_base.py [hunk 1]
```diff
@@ -61,4 +63,44 @@ def _make_pipe(self):
...
+
+        Returns
+        -------
+        dict
+            Keys are feature names and values are their proxy scores
+            Score between 0.0 and 1.0
+        """
+        proxy_scores = {}
+        
+        if method == "mutual_information":
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:90]: No need to provide a method argument since there is only one method

File: credoai/modules/dataset_modules/dataset_base.py [hunk 2]
```diff
@@ -61,4 +63,44 @@ def _make_pipe(self):
...
+            # Exclude features of type float
+            float_cols = list(self.X.select_dtypes(include=["float16", "float32", "float64"]))
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:92]: Why are we excluding `float`? You can estimate the information content between continuous and categorical variables. See [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_classif.html)

File: credoai/modules/dataset_modules/dataset_base.py [hunk 3]
```diff
@@ -61,4 +63,44 @@ def _make_pipe(self):
...
+            if excluded_columns:
+                excluded_columns.extend(float_cols)
+            else:
+                excluded_columns = float_cols
+
+            for column in self.X.drop(excluded_columns, axis=1):
+                proxy_scores[column] = normalized_mutual_info_score(
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:99]: I'd suggest we use the `feature_selection.mutual_info_clas` ([link](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_classif.html)). It handles continuous and categorical, works across the entire matrix and returns back the information for each feature.

Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:99]: Though I see that isn't `adjusted`. Hmmm

Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:99]: I still think we should use `mutual_info_classif`. It'll give a relative ordering of variables

File: credoai/modules/dataset_modules/dataset_base.py [hunk 4]
```diff
@@ -24,8 +24,10 @@ def run(self):
         pipe = self._make_pipe()
         cv_results = self._run_cv(pipe)
         group_differences = self._group_differences()
+        feature_space_proxy_scores = self._assess_proxy_in_feature_space()
         self.results = {'sensitive_feature_roc_auc': cv_results.mean(),
-                        'group_diffs': group_differences}  
+                        'group_diffs': group_differences,
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:29]: `group_differences` fails to be json serializable I believe. I'd check that.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 5]
```diff
@@ -61,4 +63,44 @@ def _make_pipe(self):
...
+                excluded_columns.extend(float_cols)
+            else:
+                excluded_columns = float_cols
+
+            for column in self.X.drop(excluded_columns, axis=1):
+                proxy_scores[column] = normalized_mutual_info_score(
+                    self.X[column], self.sensitive_features
+                )
+
+        return proxy_scores
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:103]: Once they have these scores, how do they know if something is a proxy or not? Is there a cutoff?

File: credoai/modules/dataset_modules/dataset_base.py [hunk 6]
```diff
@@ -24,8 +24,10 @@ def run(self):
         pipe = self._make_pipe()
         cv_results = self._run_cv(pipe)
         group_differences = self._group_differences()
+        feature_space_proxy_scores = self._assess_proxy_in_feature_space()
         self.results = {'sensitive_feature_roc_auc': cv_results.mean(),
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:28]:
```suggestion
        self.results = {'overall_proxy_score': cv_results.mean(),
```

Renaming to bring in line with your `feature_space_proxy_scores`

**Review by IanAtCredo — CHANGES_REQUESTED** [review]
I don't think we should be leaving out all continuous features. I offered a suggestion for this, but others may be applicable.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 2 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:92]: it does not work like that. try some combinations of this:
```
df = pd.DataFrame({'race': ['white', 'white', 'white', 'white', 'white', 'white', 'black', 'asian', 'asian', 'asian', 'asian', 'asian'],
                    'gender': ['f', 'f', 'f', 'm', 'f', 'f', 'f', 'f', 'm', 'm', 'm', 'm'],
                  'experience': [0, 0.1, 0.2, 0.4, 0.5, 0.6, 0, 0.1, 0.2, 0.4, 0.5, 0.6],
                  'mark': [0, 0.3, 0.2, 0.4, 0.5, 0.6, 0.3, 0.1, 0.2, 0.4, 0.9, 0.6],
                  'zip code': [22203, 22205, 22203, 22203, 22203, 22203, 22204, 22204, 22204, 22204, 22203, 22204]})
```

File: credoai/modules/dataset_modules/dataset_base.py (hunk 3 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:99]: Calculating mutual information between a continuous and a discrete variable is nontrivial. Only purely continuous or purely discrete pairs are possible. There is [this](https://proceedings.neurips.cc/paper/2017/file/ef72d53990bc4805684c9b61fa64a102-Paper.pdf) NIPS paper that has proposed a solution for mixed cases, but no code and looks advanced. I make updates to return mi for the two possible pairs scenarios.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 5 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:103]: Being a proxy is not a binary state.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 6 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:28]: I change this to mutual information for accuracy.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 2 above)
Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:92]: I believe the below works. You need to get the categorical variables into a coded form, and inform `mutual_info_classif`

```
df = pd.DataFrame({'race': ['white', 'white', 'white', 'white', 'white', 'white', 'black', 'asian', 'asian', 'asian', 'asian', 'asian'],
                    'gender': ['f', 'f', 'f', 'm', 'f', 'f', 'f', 'f', 'm', 'm', 'm', 'm'],
                  'experience': [0, 0.1, 0.2, 0.4, 0.5, 0.6, 0, 0.1, 0.2, 0.4, 0.5, 0.6],
                  'mark': [0, 0.3, 0.2, 0.4, 0.5, 0.6, 0.3, 0.1, 0.2, 0.4, 0.9, 0.6],
                  'zip code': [22203, 22205, 22203, 22203, 22203, 22203, 22204, 22204, 22204, 22204, 22203, 22204]})
# make longer, for sake of demo
df = pd.concat([df]*3, axis=0)

# convert cat columns to numeric to use
cat_columns = df.select_dtypes(include=['object']).columns
for col in cat_columns:
    df[col] = df[col].astype('category')
    df[f"{col}_codes"] = df[col].cat.codes

# drop categorical variables, just keep codes
processed_df = df.drop(cat_columns, axis=1)

# label which ones are categorical
cat_columns = ['zip code', 'race_codes', 'gender_codes']
discrete_features = [True if col in cat_columns else False 
                     for col in processed_df.columns]

mi= mutual_info_classif(processed_df, df['gender_codes'], 
                        discrete_features=discrete_features,random_state=42)
pd.Series(mi, index=processed_df.columns)
```

File: credoai/modules/dataset_modules/dataset_base.py (hunk 5 above)
Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:103]: I agree, but I am thinking of a user who has to decide whether to remove or keep a variable.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 2 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:92]: This returns wrong/misleading results. MI of a variable with itself must be 1.0:

experience      0.153942
mark            0.241829
zip code        0.080515
race_codes      0.245412
**gender_codes    0.679193**

Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:92]: With normalization
```
import pandas as pd
from sklearn.metrics import *
from sklearn.feature_selection import *

df = pd.DataFrame({'race': ['white', 'white', 'white', 'white', 'white', 'white', 'black', 'asian', 'asian', 'asian', 'asian', 'asian'],
                    'gender': ['f', 'f', 'f', 'm', 'f', 'f', 'f', 'f', 'm', 'm', 'm', 'm'],
                  'experience': [0, 0.1, 0.2, 0.4, 0.5, 0.6, 0, 0.1, 0.2, 0.4, 0.5, 0.6],
                  'mark': [0, 0.3, 0.2, 0.4, 0.5, 0.6, 0.3, 0.1, 0.2, 0.4, 0.9, 0.6],
                  'zip code': [22203, 22205, 22203, 22203, 22203, 22203, 22204, 22204, 22204, 22204, 22203, 22204]})
# make longer, for sake of demo
df = pd.concat([df]*3, axis=0)

# convert cat columns to numeric to use
cat_columns = df.select_dtypes(include=['object']).columns
for col in cat_columns:
    df[col] = df[col].astype('category')
    df[f"{col}_codes"] = df[col].cat.codes

# drop categorical variables, just keep codes
processed_df = df.drop(cat_columns, axis=1)

# label which ones are categorical
cat_columns = ['zip code', 'race_codes', 'gender_codes']
discrete_features = [True if col in cat_columns else False 
                     for col in processed_df.columns]

mi = mutual_info_classif(processed_df, df['gender_codes'], 
                        discrete_features=discrete_features,random_state=42)
out = pd.Series(mi, index=processed_df.columns)
normalized = out/out.max()
normalized
```

**Comment by amrasekh** [comment]
@IanAtCredo Updated the [process](https://github.com/credo-ai/credoai_lens/blob/proxy_detection/credoai/modules/dataset_modules/dataset_base.py) to return the mutual information for all possible combinations and scenarios

File: credoai/modules/dataset_modules/dataset_base.py [hunk 7]
```diff
@@ -61,4 +65,86 @@ def _make_pipe(self):
...
+        ----------
+        df : pandas.dataframe
+            A dataframe
+
+        Returns
+        -------
+        list
+            Names of categorical features
+        """        
+        cols = df.columns
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:84]: I am not comfortable with this automatic procedure.

Someone [online](https://stackoverflow.com/questions/59384802/how-can-i-automatically-detect-if-a-colum-is-categorical) proposed something like this:

```
#category 
for name, column in inputDataFrame.iteritems():
    unique_count = column.unique().shape[0]
    total_count = column.shape[0]
    if unique_count / total_count < 0.05:
        columnTypesDict[name] = 'category'
```

Essentially, they look for columns where the number of unique aspects are a low percentage of the total number. That doesn't work completely on its own either, but seems pretty good for an auto.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 8]
```diff
@@ -61,4 +65,86 @@ def _make_pipe(self):
...
+            categorical_features = self._find_categorical_features(df)
+
+        # Encode categorical features
+        for col in categorical_features:
+            df[col] = df[col].astype('category').cat.codes
+
+        discrete_features = [True if col in categorical_features else False for col in df.columns]
+
+        # Use the right mutual information methods based on the feature type of the sensitive attribute
+        if sensitive_feature_name in categorical_features:
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:127]: This is making me think we may want to have an overall randomization seed for lens that propagates through. No need to worry about that now though!

File: credoai/modules/dataset_modules/dataset_base.py (hunk 6 above)
Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:30]: This hasn't been changed yet

File: credoai/modules/dataset_modules/dataset_base.py (hunk 4 above)
Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:31]: This still needs to be addressed for the dataset assessment to work I think

**Review by IanAtCredo — CHANGES_REQUESTED** [review]

File: credoai/modules/dataset_modules/dataset_base.py (hunk 7 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:84]: I came across this too actually, but this then introduces a parameter. How to set that and explain why? 
No program can get it right at the end of the day. I added the considered feature type info to the results for the user to be aware of that. 

```
        results = {}
        for k, v in mi.items():
            if k in categorical_features:
                results[k] = {'value':v, 'feature_type': 'categorical'}
            else:
                results[k] = {'value':v, 'feature_type': 'continuous'}
```

File: credoai/modules/dataset_modules/dataset_base.py (hunk 4 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:31]: Is there anything I should do for this here?

Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:31]: You just need to make sure the output is serializable. I don't believe t currently is - you'll have to rewrite group differences. I think you may need to convert the series it outputs into a dict? Not sure.

File: credoai/modules/dataset_modules/dataset_base.py (hunk 7 above)
Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:84]: I don't think you need to reveal a parameter. It's just our way of identifying categorical variables. Better than just identifying floats I'd think. But your call

File: credoai/modules/dataset_modules/dataset_base.py (hunk 4 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:31]: done

File: credoai/modules/dataset_modules/dataset_base.py (hunk 7 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:88]: done

File: credoai/modules/dataset_modules/dataset_base.py (hunk 8 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py:138]: Sounds good!

**Review by IanAtCredo — APPROVED** [review]
